schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/users/DefaultUserService.java

```diff
@@ -156,102 +160,113 @@ private void notify(User user) {
-            db.setRealName(user.getRealName());
-
-            UserDetail model = db.getDetailRef().getModel();
-            if (model == null) {
-                Datastore.put(tx, db);
-            } else {
-                Datastore.put(tx, db, model);
-            }
-            tx.commit();
-        } finally {
-            if (tx.isActive()) tx.rollback();
-        }
         return user;
     }
 
     @Override
-    public User setPassword(User user, String newPassword) throws EntityNotFoundException {
-        Transaction tx = Datastore.beginTransaction();
-        User db = Datastore.getOrNull(tx, userMeta, user.getId());
-        if (db == null) {
-            tx.rollback();
-            throw new EntityNotFoundException();
-        }
-        db.setPassword(new ShortBlob(DigestUtil.encrypt(newPassword)));
-        Datastore.put(tx, db);
-        tx.commit();
-        return db;
+    public User setPassword(final User user, final String newPassword) throws ModelException {
+        return TransactionOperator.execute(new ModelOperation<User>() {
+            @Override
+            public User execute(com.google.appengine.api.datastore.Transaction tx) throws EntityNotFoundException {
+                User db = Datastore.getOrNull(tx, userMeta, user.getId());
+                if (db == null) {
+                    tx.rollback();
+                    throw new EntityNotFoundException();
+                }
+                db.setPassword(new ShortBlob(DigestUtil.encrypt(newPassword)));
+                Datastore.put(tx, db);
+                tx.commit();
+                return db;
+            }
+        });
     }
 
     @Override
-    public UserLogin addLogin(User user, UserLogin login) throws EntityNotFoundException, UserLoginExistsException {
+    public UserLogin addLogin(final User user, final UserLogin login) throws ModelException {
         try {
             uniqueLogin.reserve(login.getUserId(), login.getProvider());
         } catch (UniqueConstraintException e) {
             throw new UserLoginExistsException(e.getMessage());
         }
 
-        Transaction tx = Datastore.beginTransaction();
-        User db = Datastore.getOrNull(tx, userMeta, user.getId());
-        if (db == null) {
-            tx.rollback();
-            uniqueLogin.release(login.getUserId(), login.getProvider());
-            throw new EntityNotFoundException();
-        }
-        login.setId(Datastore.allocateId(db.getId(), UserLogin.class));
-        login.getUserRef().setModel(db);
+        return TransactionOperator.execute(new ModelOperation<UserLogin>() {
+            @Override
+            public UserLogin execute(com.google.appengine.api.datastore.Transaction tx) throws EntityNotFoundException {
+                User db = Datastore.getOrNull(tx, userMeta, user.getId());
+                if (db == null) {
+                    tx.rollback();
+                    uniqueLogin.release(login.getUserId(), login.getProvider());
```

Comment:
## BAD
This doesn't work...  
 1. uniqueLogin is reserved outside of TX  (If operator fails first time and works second time, the user login won't be reserved anymore)
 2. uniqueLogin creates transactions of it's own

Solution use `reserveInCurrentTransaction` and skip the whole rollback and deleting stuff